presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewInformationExtractor.java

```diff
@@ -36,25 +36,23 @@
 
 import static com.facebook.presto.sql.ExpressionUtils.removeGroupingElementPrefix;
 import static com.facebook.presto.sql.ExpressionUtils.removeSingleColumnPrefix;
-import static com.facebook.presto.sql.analyzer.SemanticErrorCode.NOT_SUPPORTED;
 import static com.google.common.base.Preconditions.checkState;
 
 public class MaterializedViewInformationExtractor
         extends DefaultTraversalVisitor<Void, Void>
 {
     private final MaterializedViewInfo materializedViewInfo = new MaterializedViewInfo();
-
     @Override
     protected Void visitQuerySpecification(QuerySpecification node, Void context)
     {
         if (node.getLimit().isPresent()) {
-            throw new SemanticException(NOT_SUPPORTED, node, "Limit clause is not supported in query optimizer");
+            materializedViewInfo.rewriteNotPossible("Limit clause is not supported in query optimizer");
```

Comment:
We should return immediately?

Reply:
This is general design problem with moving away from try/catch, and something worth discussing. Unless we add checks before every recursive call, we will inevitably be doing some unnecessary processing.

Take for example `visitQuerySpecification`:

```
            return new QuerySpecification(
                    (Select) process(node.getSelect(), context),
                    node.getFrom().map(from -> (Relation) process(from, context)),
                    node.getWhere().map(where -> (Expression) process(where, context)),
                    node.getGroupBy().map(groupBy -> (GroupBy) process(groupBy, context)),
                    node.getHaving().map(having -> (Expression) process(having, context)),
                    node.getOrderBy().map(orderBy -> (OrderBy) process(orderBy, context)),
                    node.getOffset(),
                    node.getLimit());
```
 
Even if we return early for a failure inside any one of these `process` calls, the rest of the `process` calls will still happen.

If the cost of this extra processing is nontrivial, I don't think we should make these changes at all. My understanding was that the processing was insignificant though.

Reply:
+1 on what @jainxrohit is saying... we should stop here immediately. If we feel the change is non-trivial, let's ditch it until we come up with a better design.